homeassistant/components/hydrawise/binary_sensor.py

```diff
@@ -60,8 +77,15 @@ class HydrawiseBinarySensor(HydrawiseEntity, BinarySensorEntity):
 
     def _update_attrs(self) -> None:
         """Update state attributes."""
-        if self.entity_description.key == "status":
-            self._attr_is_on = self.coordinator.last_update_success
-        elif self.entity_description.key == "is_watering":
-            assert self.zone is not None
-            self._attr_is_on = self.zone.scheduled_runs.current_run is not None
+        self._attr_is_on = getattr(self, f"_get_{self.entity_description.key}")()
+
+    def _get_status(self) -> bool:
+        return self.coordinator.last_update_success
+
+    def _get_rain_sensor(self) -> bool | None:
+        assert self.sensor is not None
+        return self.sensor.status.active
+
+    def _get_is_watering(self) -> bool:
+        assert self.zone is not None
+        return self.zone.scheduled_runs.current_run is not None
```

Comment:
You could then extend BinarySensorEntityDescription with a

```py
    is_on_fn: Callable[[HydrawiseBinarySensor], bool]
```

```py
ZONE_BINARY_SENSORS: tuple[BinarySensorEntityDescription, ...] = (
    BinarySensorEntityDescription(
        key="is_watering",
        translation_key="watering",
        device_class=BinarySensorDeviceClass.RUNNING,
        is_on_fn=lambda entity: entity.zone.scheduled_runs.current_run is not None
    ),
)
```

Then this code can become
```suggestion
        self._attr_is_on = self.entity_description.is_on_fn(self)
```

Which will move the logic of when a binary sensor should be `on` to the entity descriptions, which keeps this class with less entity specific code

Reply:
Oh, this comment was a continuation of the previous comment. If you have it like that it should work

Reply:
Now this becomes relevant to make a massive clean up :)

Reply:
Yes. It would be possible now. But I still don't like the pattern of mixing code with the sensor definitions.

Reply:
Ah, I did not know that that was your opinion. I kinda dislike that we have to extend the class when we are adding sensors. It doesn't add a lot of benefit tbf

Reply:
Ok. I updated `switch.py`, `sensor.py`, and `binary_sensor.py` to follow the `update_fn`, `turn_on_fn`, and `turn_off_fn` pattern.